### ISSUE TITLE
FileNames / TreeCommits Query times out on very large monorepo

### ISSUE THREAD

**Issue by mike-r-mclaughlin** [sourcegraph/sourcegraph#41151]
- **Sourcegraph version:** 3.42.0
- **Reported by:** https://github.com/sourcegraph/accounts/issues/6716 [via Slack](https://sourcegraph.slack.com/archives/C01D50MSA7L/p1661840779425309)

#### Description

On a large monorepo, we see a large number of 504 errors.  This monorepo has 2M+ files in it.

When investigating, it appears to all be from 1 or 2 graphql APIs:

##### FileNames (~90%)

Example Query:
```
query FileNames($repository: String!, $commit: String!) {
  repository(name: $repository) {
    id
    commit(rev: $commit) {
      id      
      fileNames
      __typename
    }
    __typename
  }
}
```

```
Variables
{
    "repository": "hostname/path/to/monorepo",
    "commit": ""
}
```

#### Steps to reproduce:

1. Open up the monorepo in Sourcegraph
2. Wait

#### Expected behavior:

The files tab either loads on demand or loads quickly.

#### Actual behavior:

The files tab and/or the view of directories in the repo is slow and/or fails to load.

/cc @sourcegraph/search-core

**Comment by jjeffwarner** [comment]
@mike-r-mclaughlin Have they run the git maintenance command on this repo recently?

**Comment by keegancsmith** [comment]
TL;DR FileNames returns all filenames in a repo at a commit. For gigarepo (3.7M files) this takes 26s and is 15MB compressed (380MB uncompressed). Best guess is the overhead of subrepo perms is making this timeout. Next steps is reproducing and then hopefully improving perf here.

This query is for the FuzzyFinder (ctrl-k shortcut). It shouldn't affect the files side bar or directory view. The FuzzyFinder is also designed to handle this being slow / failing.

Did the customer actually experience the files sidebar not loading or directory view not loading? Can they share a screenshot? This might be TreeCommit failing, but I didn't dive into that yet.

I was experimenting with our gigarepo, and I believe I noticed slowness in my browser. I didn't dive deeper, but given the numbers below I wouldn't be surprised if it was due to the fuzzy finder. So if we have had any complaints about slowness in the browser from the customer, I would suspect this. Note: gigarepo is bigger, it has about 3.7M files.

It took about 26s for the query to finish. The raw ls-files time is below, and is about 10s with a warm file cache. So there is a lot of overhead in our API around JSON marshalling/etc I assume. But given the size, I imagine this should be working for the customer. However, the missing customer is using sub repo permissions, and that isn't part of my dogfood testing.

@mollylogue can you point me in the direction of a representative rules file? I would like to do some performance testing / optimization of it for LsFiles.

<details>
```
gitserver-3:/data/repos/gigarepo$ git ls-files -z --with-tree HEAD | wc -c
382964441
gitserver-3:/data/repos/gigarepo$ git ls-files -z --with-tree HEAD | gzip - | wc -c
15644094
gitserver-3:/data/repos/gigarepo$ time git ls-files -z --with-tree HEAD > /dev/null
real	0m 10.11s
user	0m 7.65s
sys	0m 2.28s
gitserver-3:/data/repos/gigarepo$ sh git-stats
21.6G	. gitdir
1943168 commits

HEAD statistics
17.487GiB
3697524 files
histogram:
10^0	7765
10^1	5540
10^2	50116
10^3	2121429
10^4	1333493
10^5	161430
10^6	16187
10^7	1482
10^8	82
```
</details>

**Comment by keegancsmith** [comment]
I should of commented earlier, but TreeCommit is us showing recent commits for a path. This is the known slow endpoint for monorepos (IE it is like blame, it has to look at lots of commits)

**Comment by mike-r-mclaughlin** [comment]
> Did the customer actually experience the files sidebar not loading or directory view not loading? Can they share a screenshot? 

This is something I forgot to mention.  The files sidebar is loaded and shown well in advance of the 504 (or the request finishing if it finishes prior to 60s).  It's not clear if the browser is caching it from a previous release or something else.  When looking at all other requests, no other request had a response that was close to the size of the `FileNames` API request (9mb compressed).  So, it must be getting that from cache.  However, we did have the browser cache disabled at the time (via browser developer tools).

We watched the network tab in the debug window and saw it take anywhere from 25s to over 60s (504 / timeout).

When `TreeCommits` is slow, the UI shows a spinner.  This is much less frequent than the `FileNames` behavior.

**Comment by keegancsmith** [comment]
Can I get an understanding of prioritization here? `FileNames` API failing _only_ affects the fuzzy finder, which I assume most users don't use. Otherwise it is spam on the monitoring.

`TreeCommits` is also an "optional" part of the UI, and not as core. So would be great to get an understanding of the actual reported user impact.

Either way this issue is on my plate, its just a question of how much time I invest in it and the priority vs other tasks.

> This is something I forgot to mention. The files sidebar is loaded and shown well in advance of the 504 (or the request finishing if it finishes prior to 60s). It's not clear if the browser is caching it from a previous release or something else.

`FileNames` API request is not used for the files sidebar. It is _only_ used for the fuzzy finder (accessed with cmd-k shortcut).

> We watched the network tab in the debug window and saw it take anywhere from 25s to over 60s (504 / timeout).

After what I saw on dogfood I very strongly suspect the overhead is sub repo permissions. I took a look at the code, and there are wins around orientating permission checking to be optimized for batch operations. Right now there is a fair amount of overhead we do per _filename_, so the wins here are amortizing a bunch of that work so we don't pay it per filename. There are more algorithmic/datastructure wins we could also pursue but those are quite complicated.

However, I have not yet run a profiler with subrepo perms enabled (will do so once I have a representative rules files). So the above is all based on intuition and reading code (versus actually measuring).

> When `TreeCommits` is slow, the UI shows a spinner. This is much less frequent than the `FileNames` behavior.

How important is this problem?

**Comment by keegancsmith** [comment]
Added a benchmark for batch file filtering https://github.com/sourcegraph/sourcegraph/pull/41257 no perf work yet on it.

**Comment by keegancsmith** [comment]
First fruit from the benchmark. A 1250% speed up of sub repo permissions (in that benchmark). On my computer the benchmark goes from processing 95k paths per second to 1.2m paths per second. https://github.com/sourcegraph/sourcegraph/pull/41258

**Comment by mike-r-mclaughlin** [comment]
> Can I get an understanding of prioritization here? FileNames API failing only affects the fuzzy finder, which I assume most users don't use. Otherwise it is spam on the monitoring.

> FileNames API request is not used for the files sidebar. It is only used for the fuzzy finder (accessed with cmd-k shortcut).

Ah OK.  That makes sense.  

The issue here is that we're getting 504 errors because of this (`FileNames`) API timing out.  And, this customer's admin team is being measured on the "error rate" (i.e., they have a dashboard for every enterprise app that shows number of requests vs number of requests with errors).  And, because there are so many 504's, it is setting off alarms.  This also only started happening after 3.42.0 upgrade.  So, there was concerns that it was something to do with 3.42.0.  But, an important note, we fixed a few sub-repo permissions issues that would allow users to see more files in 3.42.0 so that is likely the root cause to seeing more 504s after 3.42.0.

In addition, we're seeing slow performance at various places in the UI and as we troubleshoot this, we've been trying to file an issue for each one.  This is just another example.  

If there is a relatively easy fix to keep this from loading every single time the page loads (i.e., could we lazily load this data only when the user first tries to use the fuzzy finder?), then I would say let's do that quick fix now.  Otherwise, I'm fine with lowering the priority while we make a full list of each slow performance item.


> How important is this [TreeCommits] problem?

It's not all that important if it doesn't impact the user experience.  When the UI "hangs" (and we see the spinner), it usually appears to be because of this API.  That's just from watching the developer networking tab and the UI at the same time.   It may be something else entirely.  Let's separate these into 2 separate tickets.  And, I'll get a better screenshot / video recording of the problem.
